### PR TITLE
Support a 'recommended' completion entry

### LINKED ISSUE #19855: [typescript] Offer IntelliSense suggestions based on method parameter or local type
_From @rstarkov on November 7, 2017 12:25_

It would be useful if IntelliSense offered suggestions based on the expected type of the expression being edited. Consider the following examples:

    function Foo(thing: Namespace.Stuff.SomeClass) { ... }

    Foo(new |
            ^ here

    let x: Namespace.Stuff.SomeClass;
    x = new |
            ^ here

In these cases, "Namespace.Stuff.SomeClass" should be offered by IntelliSense at the indicated positions.

This behaviour would also be handy for enums: the enum type could be offered in the same contexts as above (without the `new` keyword obviously).

For enums there's another scenario where this would be handy, namely when I'm comparing an enum-typed local or parameter - IntelliSense should offer the enum type for autocompletion after I type the comparison operator.

_Copied from original issue: Microsoft/vscode#37771_

### PR DESCRIPTION
**Pull request by andy-ms**
Fixes #19855
This adds an `isRecommended` boolean to one (or no) completion entry, when that entry is for a enum or (non-abstract) class and the context indicates that that's the only allowed type.

### CODE REVIEW

File: src/services/completions.ts [hunk 1]
```diff
@@ -160,10 +161,20 @@ namespace ts.Completions {
...
+            isRecommended: trueOrUndef(isRecommendedCompletionMatch(symbol, recommendedCompletion, typeChecker)),
         };
     }
 
+    function isRecommendedCompletionMatch(localSymbol: Symbol, recommendedCompletion: Symbol, checker: TypeChecker): boolean {
+        return localSymbol === recommendedCompletion ||
+            !!(localSymbol.flags & SymbolFlags.ExportValue) && checker.getExportSymbolOfSymbol(localSymbol) === recommendedCompletion;
+    }
+
+    function trueOrUndef(b: boolean): true | undefined {
```
Comment (armanio123) [inline src/services/completions.ts]: Avoid abbreviations.

File: src/services/completions.ts [hunk 2]
```diff
@@ -524,9 +536,32 @@ namespace ts.Completions {
...
+        const symbol = ty && ty.symbol;
+        // Don't include make a recommended completion for an abstract class
+        return symbol && (symbol.flags & SymbolFlags.Enum || symbol.flags & SymbolFlags.Class && !isAbstractConstructorSymbol(symbol))
+            ? getFirstSymbolInChain(symbol, currentToken, checker)
+            : undefined;
+    }
+
+    function getFirstSymbolInChain(symbol: Symbol, enclosingDeclaration: Node, checker: TypeChecker): Symbol | undefined {
+        const chain = checker.getAccessibleSymbolChain(symbol, enclosingDeclaration, /*meaning*/ SymbolFlags.All, /*useOnlyExternalAliasing*/ false);
+        return chain
```
Comment (armanio123) [inline src/services/completions.ts]: Nested shorthands are hard to read. Consider if else, or switch statement.

**Comment by andy-ms** [comment]
Should probably get review from @mjbvz

**Comment by mjbvz** [comment]
Just to confirm, the expected UX here would be that we surface the suggested completion first? Should that also happen if the user has already started typing something? Would sort order be enough here then?

File: src/compiler/checker.ts [hunk 3]
```diff
@@ -13876,6 +13859,12 @@ namespace ts {
                 case SyntaxKind.AmpersandAmpersandToken:
                 case SyntaxKind.CommaToken:
                     return node === right ? getContextualType(binaryExpression) : undefined;
+                case SyntaxKind.EqualsEqualsEqualsToken:
```
Comment (mhegazy) [inline src/compiler/checker.ts]: this changes what contextual type mean in the language. we should talk about this first

File: src/compiler/checker.ts [hunk 4]
```diff
@@ -14091,9 +14080,13 @@ namespace ts {
                     return getContextualTypeForReturnExpression(node);
                 case SyntaxKind.YieldExpression:
                     return getContextualTypeForYieldOperand(<YieldExpression>parent);
-                case SyntaxKind.CallExpression:
                 case SyntaxKind.NewExpression:
-                    return getContextualTypeForArgument(<CallExpression>parent, node);
+                    if (node.kind === SyntaxKind.NewKeyword) { // for completions after `new `
```
Comment (mhegazy) [inline src/compiler/checker.ts]: this one too

File: src/services/types.ts [hunk 5]
```diff
@@ -712,19 +712,16 @@ namespace ts {
...
     }
 
+    // see comments in protocol.ts
     export interface CompletionEntry {
         name: string;
         kind: ScriptElementKind;
-        kindModifiers: string;   // see ScriptElementKindModifier, comma separated
+        kindModifiers: string;
         sortText: string;
-        /**
```
Comment (mhegazy) [inline src/services/types.ts]: why is this comment removed

File: src/services/types.ts [hunk 6]
```diff
@@ -712,19 +712,16 @@ namespace ts {
...
-         * An optional span that indicates the text to be replaced by this completion item. It will be
-         * set if the required span differs from the one generated by the default replacement behavior and should
-         * be used in that case
-         */
         replacementSpan?: TextSpan;
         hasAction?: true;
         source?: string;
+        isRecommended?: true;
```
Comment (mhegazy) [inline src/services/types.ts]: i would rather we put this on the CompletionInfo instead. i understand for the server we have to put it on the entry.. but for VS for example, it makes more sense to have it on the info..
Please check with @amcasey about this too.

**Comment by mhegazy** [comment]
> Just to confirm, the expected UX here would be that we surface the recommended completion first? Should that also happen if the user has already started typing something? Would sort order be enough here then?

Visual Studio will show the recommended completion as selected if nothing is typed. the order does not change. I would recommend doing the same for VSCode; changing the order is disorienting.

animation

File: src/services/types.ts (hunk 6 above)
Comment (amcasey) [reply src/services/types.ts]: I'm not intimately familiar with these types.  The info is the wrapper object containing a list of entries?  Assuming that's the case, having it on the entry is the most convenient place for VS.

Comment (andy-ms) [reply src/services/types.ts]: It's defined at `interface CompletionInfo` in `services/types.ts`

Comment (amcasey) [reply src/services/types.ts]: Then, yes, the entry would be preferable to the info.  I've created a PR with the VS change if you'd like to test it.

**Comment by andy-ms** [comment]
@amcasey Good to go?

**Comment by amcasey** [comment]
So far, it's not working.

```ts
console.log('Hello world');

class CA { }
class CB { }

const c: CB = //caret
```

 1. If I invoke completion there, the recommended completion is "CB".  I actually wanted "new".
 2. "CB" isn't pre-selected.  I'm still investigating why not.

**Comment by amcasey** [comment]
I've confirmed that the flag is getting to Roslyn, but I have yet to figure out why it's having no effect.

**Comment by amcasey** [comment]
Looks like a problem in the managed component.  I think you're good to go.

**Comment by andy-ms** [comment]
@amcasey Thanks!

**Pull request merged**